continuation of "support recursive bwrap" : #172

giuseppe opened the pull request:
continuation of https://github.com/projectatomic/bubblewrap/pull/172

I've added a fixup patch and a patch to handle recursive pid namespaces to mount`/proc` recursively

giuseppe commented:
any comment on this PR?

rhatdan commented:
LGTM
@cgwalters @alexlarsson PTAL

giuseppe commented:
ping?

cgwalters commented:
Sorry give me a bit to bubble this up to the top of my queue, so to speak...I assume alexl is busy too.  Would definitely like this.

cgwalters commented on bubblewrap.c:
```diff
@@ -1020,17 +1020,20 @@ setup_newroot (bool unshare_pid,
-          const char *cover_proc_dirs[] = { "sys", "sysrq-trigger", "irq", "bus" };
-          for (i = 0; i < N_ELEMENTS (cover_proc_dirs); i++)
+             user, but lets cover them anyway if we have no CAP_SYS_ADMIN just to make sure */
+          if (!opt_cap_add_or_drop_used || !(requested_caps[0] & CAP_TO_MASK_0 (CAP_SYS_ADMIN)))
```
Hmm.  We don't need to cover them if they're *already* read-only right?  Would that help the logic here?

giuseppe replied:
I've tried this approach in the fixup patch I've added.  It should work as well:

previously:
```
$ bwrap --bind / / --proc /proc findmnt -U --noheadings -o TARGET -R /proc
/proc
├─/proc/sys
│ └─/proc/sys/fs/binfmt_misc
├─/proc/sysrq-trigger
├─/proc/irq
└─/proc/bus
```

with the PR:
```
$ ./bwrap --bind / / --proc /proc findmnt -U --noheadings -o TARGET -R /proc
/proc
└─/proc/sys/fs/binfmt_misc
```

alexlarsson commented on bubblewrap.c:
```diff
@@ -2351,6 +2351,40 @@ main (int    argc,
+      die_with_error ("setting up newroot bind");
+    if (chdir ("/newroot") != 0)
+      die_with_error ("chdir /newroot");
+    if (mkdtemp (pivot_tmp) == NULL)
```
I don't like how this writes to the new root. For example, this will not really work if you do a `--ro-bind / /`, or if the root is a read-only container image. Testing in `tmp` too makes it a bit more likely that it would work, but there is no guarantees that there is a tmp, or that it is writable. I really would like us to have minimal requiements for the target filesystem layout.

giuseppe replied:
barely tested but would something like the fixup be enough?

Also, the original patch is coming from @cgwalters.  Does something like in the fixup look good to you?

cgwalters replied:
> For example, this will not really work if you do a --ro-bind / /, or if the root is a read-only container image. 

I use the  `--ro-bind / /` pattern a lot myself for quick sanity checking of bwrap, but it feels like we really want a `--bind-host` or something? 

Are there really use cases for spawning a container with no writable filesystems?

Actually, couldn't we special case this and detect the case where the user is binding `/` and skip doing the tmpfs etc. entirely?

giuseppe replied:
the last patch I've added drops the requirement for write access to the destination root:

```
./bwrap --ro-bind / / --proc /proc ./bwrap --ro-bind / / echo hello
hello
```

alexlarsson commented:
I really like this approach, because it doesn't need any weird persistent files to be created, its all owned by cwd and open fds. However, it does seem to be against what the manpage says should work, so at the very least we should try it on a bunch of kernels and see which ones it works on, and ideally talk to some kernel people to ensure its not something randomly working that could break later.

giuseppe commented:
yes it is not a documented behaviour and I've firstly seen it used in runc:

https://github.com/opencontainers/runc/blob/master/libcontainer/rootfs_linux.go#L647-L674

I can test it on other kernels if you wish, although it seems every Docker container out there is already using it :slightly_smiling_face:

alexlarsson commented:
If runc uses it then i'm more confident, but it would still be nice to know that this works for kernels of interest like debian stable, rhel7, ubuntu lts, etc.

giuseppe commented:
here the results:
```
Debian 9:
# ./bwrap --ro-bind / / --proc /proc ./bwrap --ro-bind / / cat /etc/os-release
PRETTY_NAME="Debian GNU/Linux 9 (stretch)"
NAME="Debian GNU/Linux"
VERSION_ID="9"
VERSION="9 (stretch)"
ID=debian
HOME_URL="https://www.debian.org/"
SUPPORT_URL="https://www.debian.org/support"
BUG_REPORT_URL="https://bugs.debian.org/"

Ubuntu 16.04:
# ./bwrap --ro-bind / / --proc /proc ./bwrap --ro-bind / / cat /etc/os-release
NAME="Ubuntu"
VERSION="16.04.4 LTS (Xenial Xerus)"
ID=ubuntu
ID_LIKE=debian
PRETTY_NAME="Ubuntu 16.04.4 LTS"
VERSION_ID="16.04"
HOME_URL="http://www.ubuntu.com/"
SUPPORT_URL="http://help.ubuntu.com/"
BUG_REPORT_URL="http://bugs.launchpad.net/ubuntu/"
VERSION_CODENAME=xenial
UBUNTU_CODENAME=xenial

CentOS 7:
# ./bwrap --ro-bind / / --proc /proc ./bwrap --ro-bind / / cat /etc/os-release
NAME="CentOS Linux"
VERSION="7 (Core)"
ID="centos"
ID_LIKE="rhel fedora"
VERSION_ID="7"
PRETTY_NAME="CentOS Linux 7 (Core)"
ANSI_COLOR="0;31"
CPE_NAME="cpe:/o:centos:centos:7"
HOME_URL="https://www.centos.org/"
BUG_REPORT_URL="https://bugs.centos.org/"

CENTOS_MANTISBT_PROJECT="CentOS-7"
CENTOS_MANTISBT_PROJECT_VERSION="7"
REDHAT_SUPPORT_PRODUCT="centos"
REDHAT_SUPPORT_PRODUCT_VERSION="7"
```

giuseppe commented on bubblewrap.c:
```diff
@@ -2357,6 +2363,29 @@ main (int    argc,
+  { cleanup_fd int oldrootfd = open ("/", O_DIRECTORY | O_RDONLY);
+    if (oldrootfd < 0)
+      die_with_error ("can't open /");
+    if (mount ("/newroot", "/newroot", NULL, MS_MGC_VAL | MS_BIND | MS_REC, NULL) < 0)
```
I think we can drop it completely as it seems the `pivot_root(".", ".")` doesn't require newroot and oldroot to be on different file systems, which I guess was the reason it was added by the Colin's patch.

@cgwalters are you fine if I drop it?

giuseppe commented on bubblewrap.c:
```diff
@@ -2357,6 +2363,29 @@ main (int    argc,
+      die_with_error ("setting up newroot bind");
+    if (chdir ("/newroot") != 0)
+      die_with_error ("chdir /newroot");
+    if (pivot_root (".", ".") != 0)
```
comment added.

giuseppe commented:
added a fixup patch to address the last comments.  I've dropped the additional mount point, but let's wait for @cgwalters to review the change as he originally added it.

alexlarsson commented:
Oh, i didn't mean to drop the mount point for /newroot. I can imagine that being useful because things expect / to be a mountpoint. However, you can mount it before you mount anything on top of it, right when you mkdir it.

giuseppe commented:
ah I see now, that can be a problem if there is no bind mount specified for `/'.

I've added another fixup to re-introduce the bind mount just after the `mkdir`

alexlarsson commented:
That looks good to me, @cgwalters ?

cgwalters commented on tests/test-run.sh:
```diff
@@ -113,6 +113,15 @@ $RUN --unshare-pid --as-pid-1 --bind / / bash -c 'echo $$' > as_pid_1.txt
+if ! test -u ${BWRAP}; then
+    echo "ok - # SKIP no --cap-add support"
+else
+    $BWRAP --unshare-all --uid 0 --gid 0 --cap-add ALL --bind / / --proc /proc \
```
One issue with all of our tests is we're using `--bind / /` which IMO is an unusual/weird case.   Also you seem to be skipping these tests in the non-setuid case - why?
 
I pushed a fixup for both of these.

giuseppe replied:
we are skipping them in the non-setuid case as bwrap will complain that only root can use `cap-add`

giuseppe replied:
are you fine if we add again the check for non-setuid case?

cgwalters replied:
It's not about setuid but about support for user namespaces....which is currently sort of the same thing because if the distro has user namespaces, there's no reason to install bwrap setuid.

To say this another way, it's fine to use `--cap-add ALL` as non-root if `--unshare-user`, right?

On Fedora we definitely shouldn't be installing setuid, and the tests definitely work for me currently.  We may however need the suid check on CentOS7 today.

cgwalters replied:
Found the bug where we were mistakenly doing `chmod u+s` on Fedora.

cgwalters commented:
OK, @giuseppe ^

giuseppe commented:
LGTM!

cgwalters commented:
@rh-atomic-bot r+ 52b3272
